index/scorch/scorch.go

```diff
@@ -387,25 +391,56 @@ func (s *Scorch) Batch(batch *index.Batch) (err error) {
 		ids = append(ids, docID)
 	}
 
-	// FIXME could sort ids list concurrent with analysis?
+	var synonymDocuments = make([]*index.SynonymDefinition, len(batch.SynOps))
+	synDocIndex := 0
+	for docID, doc := range batch.SynOps {
+		if doc != nil {
+			// insert _id field
+			doc.AddIDField()
+			numSynonyms++
+			synonymDocuments[synDocIndex] = doc.SynonymInfo()
+			synDocIndex++
+			numPlainTextBytes += doc.NumPlainTextBytes()
+		} else {
+			numDeletes++
+		}
+		ids = append(ids, docID)
+	}
+	synonymDocuments = synonymDocuments[:synDocIndex]
+	var synNameToAnalyzedDocs sync.Map
 
+	// FIXME could sort ids list concurrent with analysis?
 	if numUpdates > 0 {
-		go func() {
-			for k := range batch.IndexOps {
-				doc := batch.IndexOps[k]
-				if doc != nil {
-					// put the work on the queue
-					s.analysisQueue.Queue(func() {
-						analyze(doc, s.setSpatialAnalyzerPlugin)
-						resultChan <- doc
-					})
+		if numSynonyms > 0 {
```

Comment:
bro tht analyze function is called elsewhere also so there i would have to send nil vals for synonym documents + like if an existing user has not set any synonym docs and is doing normal indexing, then calling analyzeWithSynonym will needlessly involve an unnecessary function call to analyzeSynonym so that is why i kept it different

Reply:
passing nil value in case of non-synonym specific cases is fine in my opinion but it can be guarded against analyzeSynonym by seeing if there are no set of synonym documents passed to the batch right? 

that said, there is still the fact that a scorch.Analyze() function exists which passes the document it was given to analyze (and shouldn't do a synonym specific analysis which would be handled when you pass a nil value). I think this behaviour would be fine, however I'd suggest getting @abhinavdangeti opinion on this as well, since the analyze() function is very close to the user level.